sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterName.java

```diff
@@ -0,0 +1,159 @@
+ */
+package com.google.cloud.dataflow.sdk.util.common;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ * A structured counter name.
+ *
+ * <p>It holds all information for generating structured name protos.
+ * It is also used as a wrapper during the migration to structured counters.
+ *
+ * <p>Before migration, the {@link CounterName} will be converted to a flat name.
+ * During migration, the {@link CounterName} will be mapped to a structured name proto.
+ */
+public class CounterName {
+  /**
+   * Returns a {@link CounterName} with the given name.
+   */
+  public static CounterName named(String name) {
+    return new CounterName(name, "", "", "");
+  }
+
+  /**
+   * Returns a msecs {@link CounterName}.
+   */
+  public static CounterName msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given origin.
+   */
+  public CounterName withOrigin(String origin) {
+    return new CounterName(this.name, origin, this.stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given step name.
+   */
+  public CounterName withStepName(String stepName) {
+    return new CounterName(this.name, this.origin, stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given prefix.
+   */
+  public CounterName withPrefix(String prefix) {
+    return new CounterName(this.name, this.origin, this.stepName, prefix);
+  }
+
+  /**
+   * Name of the counter.
+   *
+   * <p>For example, process-msecs, ElementCount.
+   */
+  private final String name;
+
+  /**
+   * Origin (namespace) of counter name.
+   *
+   * <p>For example, "user" for user-defined counters.
+   * It is empty for counters defined by the Dataflow service or SDK.
+   */
+  private final String origin;
+
+  /**
+   * System defined step name.
+   *
+   * <p>For example, s1, s2.out.
+   * It may be empty when counters don't associate with step names.
+   */
+  private final String stepName;
+
+  /**
+   * Prefix of group of counters.
+   *
+   * <p>It is empty when counters don't have general prefixes.
+   */
+  private final String prefix;
+
+  /**
+   * Flat name is the equivalent unstructured name.
+   *
+   * <p>It is null before {@link #getFlatName()} is called.
+   */
+  @Nullable
+  private String flatName;
+
+  private CounterName(String name, String origin, String stepName, String prefix) {
+    this.name = checkNotNull(name, "name");
+    this.origin = checkNotNull(origin, "origin");
+    this.stepName = checkNotNull(stepName, "stepName");
+    this.prefix = checkNotNull(prefix, "prefix");
+    this.flatName = null;
+  }
+
+  /**
+   * Returns the flat name of a structured counter.
+   */
+  public synchronized String getFlatName() {
+    if (flatName == null) {
+      synchronized (this) {
```

Comment:
This still has `synchronized` on getFlatName(). I'm also not convinced we need synchronized at all.

We could do this with an AtomicReference for the flatName. Essentiallly:

```
String name = flatName.get();
if (name == null) {
   StringBuilder sb = ...;
   flatName.compareAndSet(null, sb.build());
   name = flatName.get();
}
return name;
```

Not sure the how badly we want this to be lock free.